Source/Utils/ParticleUtils.H

```diff
@@ -170,6 +170,190 @@ namespace ParticleUtils {
+            const amrex::ParticleReal ptdotu = (pxtot*ux + pytot*uy + pztot*uz);
+
+            // Compute coefficients for quadratic equation for alpha.
+            const amrex::ParticleReal a = A*A*dbetasq - ptdotu*ptdotu;
+            const amrex::ParticleReal b = D*ptdotu - 2.0_prt*A*A*p2dotu;
+            const amrex::ParticleReal c = A*A*E2*E2 - D*D/4.0_prt;
+
+            const amrex::ParticleReal root = b*b - 4.0_prt*a*c;
+            if (root < 0.0 || a == 0.0) { continue; }
+            const amrex::ParticleReal alpha = (-b + std::sqrt(root))/(2.0_prt*a);
+
+            // Update particle velocities.
+            const amrex::ParticleReal ratio1 = alpha/(wpmp1*c2);
+            const amrex::ParticleReal ratio2 = alpha/(wpmp2*c2);
+            uxp[ indices[i1] ] += ratio1*ux;
+            uyp[ indices[i1] ] += ratio1*uy;
+            uzp[ indices[i1] ] += ratio1*uz;
+            uxp[ indices[i2] ] -= ratio2*ux;
+            uyp[ indices[i2] ] -= ratio2*uy;
+            uzp[ indices[i2] ] -= ratio2*uz;
+
+            Erel_cumm += Erel - deltaEpair;
+
+            if (i1 < cell_stop - 2) {
+                // If not done with the loop, increment to the next pair of particles
+                i1 = i2 + 1;
+            } else {
+                loop_count++;
+                // On next time through the loop, use a different set of pairs.
+                // What index to start with depends on whether the number of particles
+                // is even or odd.
+                if (i1 == cell_stop - 2) {
+                    if ((numCell % 2) == 0) { i1 = cell_start + 1; }
+                    if ((numCell % 2) != 0) { i1 = cell_start; }
+                } else {
+                    if ((numCell % 2) == 0) { i1 = cell_start; }
+                    if ((numCell % 2) != 0) { i1 = cell_start + 1; }
+                }
+                // Compute the energy fraction to correct with one more loop.
+                // This is the remaining energy correction divided by the cummulated Erel,
+                // the amount of "free" energy available.
+                // If deltaE < 0, this check always passes since the amount of energy that
+                // can be added is unlimited.
+                energy_frac_eff = std::abs(deltaE)/Erel_cumm;
+                if (energy_frac_eff > energy_fraction_max || loop_count > max_loop_count) {
+#ifndef AMREX_USE_GPU
+                    if (numCell > 50) {
+                        amrex::Print() << "Notice: energy_frac_eff = " << energy_frac_eff << "\n";
+                        amrex::Print() << "        on loop_count = " << loop_count << "\n";
+                        amrex::Print() << "        is larger than energy_frac_max = " << energy_fraction_max << "\n";
+                        amrex::Print() << "        num_parts    = " << numCell << "\n";
+                        amrex::Print() << "        deltaE     = " << deltaE << "\n";
+                    }
+#endif
```

Comment:
I'm guessing these print statements are just for debugging and will be removed, right?

Reply:
This is now cleaned up.